users/talk.rs

```diff
@@ -0,0 +1,1240 @@
+    // Handle user input from stdin and send it to the write stream.
+    handle_stdin_input(write_stream, split_row, top_line)?;
+
+    Ok(())
+}
+
+// Spawns a new thread to read data from the TCP stream and display it on the terminal.
+fn spawn_read_thread(
+    stream: TcpStream,
+    split_row: u16,
+    width: u16,
+    top_line: Arc<Mutex<u16>>,
+    bottom_line: Arc<Mutex<u16>>,
+) -> Result<(), io::Error> {
+    thread::spawn(move || {
+        // Attempt to draw the terminal and handle any errors
+        let mut handle = match draw_terminal(split_row, width) {
+            Ok(handle) => handle,
+            Err(e) => {
+                eprintln!("Failed to draw terminal: {}", e);
+                return;
+            }
+        };
+
+        let mut buffer = [0; 128];
+        let mut stream = stream;
+
+        loop {
+            match stream.read(&mut buffer) {
+                Ok(nbytes) => {
+                    if nbytes > 0 {
+                        // Handle user input and manage potential errors
+                        if let Err(e) = handle_user_input(
+                            &mut handle,
+                            std::str::from_utf8(&buffer[..nbytes]).unwrap_or(""),
+                            split_row,
+                            Arc::clone(&top_line),
+                            Arc::clone(&bottom_line),
+                        ) {
+                            eprintln!("Error handling user input: {}", e);
+                            break;
+                        }
+                    } else {
+                        // Handle connection closure
+                        handle_connection_close();
+                        break;
+                    }
+                }
+                Err(e) => {
+                    eprintln!("Error reading from stream: {}", e);
+                    break;
+                }
+            }
+        }
+    });
+
+    Ok(())
+}
+
+// Handles user input and displays it in the bottom section of the terminal.
+fn handle_user_input(
+    handle: &mut io::StdoutLock,
+    input: &str,
+    split_row: u16,
+    top_line: Arc<Mutex<u16>>,
+    bottom_line: Arc<Mutex<u16>>,
+) -> io::Result<()> {
+    let top_line = top_line.lock().unwrap();
+    let mut bottom_line = bottom_line.lock().unwrap();
+
+    // Reset bottom line position if it exceeds the available space.
+    if *bottom_line >= split_row - 2 {
+        *bottom_line = 0;
+    }
+
+    // Move cursor to bottom window
+    write!(handle, "\x1b[{};0H", split_row + *bottom_line + 1)?;
+    writeln!(handle, "{}", input)?;
+
+    // Move cursor to top window
+    write!(handle, "\x1b[{};H", *top_line)?;
+
+    handle.flush()?;
+
+    *bottom_line += 1;
+
+    Ok(())
+}
+
+// Draws the terminal interface with a split line separating the top and bottom sections.
+fn draw_terminal(split_row: u16, width: u16) -> io::Result<io::StdoutLock<'static>> {
```

Comment:
Probably best to switch to the raw mode too. I haven't seen you doing it anywhere. I expect there's quite a bit of lag of the input while in cooked mode.